lib/licensed/source/bundler.rb

```diff
@@ -35,14 +36,133 @@ def dependencies
+
+        # delete any specifications loaded from a gemspec
+        all_specs.delete_if { |s| s.source.is_a?(::Bundler::Source::Gemspec) }
+      end
+
+      # Recursively finds the dependencies for Gem specifications.
+      # Returns a `Set` containing the package names for all dependencies
+      def recursive_specs(specs, results = Set.new)
+        return [] if specs.nil? || specs.empty?
+
+        new_specs = Set.new(specs) - results.to_a
+        return [] if new_specs.empty?
+
+        results.merge new_specs
+
+        dependency_specs = new_specs.flat_map { |s| specs_for_dependencies(s.dependencies, s.source) }
+
+        return results if dependency_specs.empty?
+
+        results.merge recursive_specs(dependency_specs, results)
+      end
+
+      # Returns the specs for dependencies that pass the checks in `include?`
+      # Raises an error if the specification isn't found
+      def specs_for_dependencies(dependencies, source)
+        included_dependencies = dependencies.select { |d| include?(d, source) }
+        included_dependencies.map do |dep|
+          gem_spec(dep) || raise("Unable to find a specification for #{dep.name} (#{dep.requirement}) in any sources")
+        end
+      end
+
+      # Returns a Gem::Specification for the provided gem argument.  If a
+      # Gem::Specification isn't found, an error will be raised.
+      def gem_spec(dependency)
+        return unless dependency
+
+        # bundler specifications aren't put in ::Bundler.specs_path, even if the
+        # gem is a runtime dependency.  it needs to be handled specially
+        return bundler_spec if dependency.name == "bundler"
+
+        # find a specifiction from the resolved ::Bundler::Definition specs
+        spec = definition.resolve.find { |s| s.satisfies?(dependency) }
+        return spec unless spec.is_a?(::Bundler::LazySpecification)
+
+        # if the specification is coming from a gemspec source,
+        # we can get a non-lazy specification straight from the source
+        if spec.source.is_a?(::Bundler::Source::Gemspec)
+          return spec.source.specs.first
+        end
+
+        # look for a specification at the bundler specs path
+        spec_path = ::Bundler.specs_path.join("#{spec.full_name}.gemspec")
+        return unless File.exist?(spec_path.to_s)
+        Gem::Specification.load(spec_path.to_s)
+      end
+
+      # Returns whether a dependency should be included in the final
+      def include?(dependency, source)
+        # ::Bundler::Dependency has an extra `should_include?`
+        return false unless dependency.should_include? if dependency.respond_to?(:should_include?)
+
+        # Don't return gems added from `add_development_dependency` in a gemspec
+        # if the :development group is excluded
+        gemspec_source = source.is_a?(::Bundler::Source::Gemspec)
+        return false if dependency.type == :development && (!gemspec_source || exclude_development_dependencies?)
```

Comment:
This filters out gems that are added to a gemspec via `add_development_dependency`.  `!gemspec_source` covers dependencies that are found recursively from other dependencies (`source` should be `nil`).  `exclude_development_dependencies?` checks that the `:development` group is excluded via bundler or licensed configuration.

If bundler was listed as a development dependency via a local gemspec, or if it was found recursively as a development dependency of a runtime dependency, it will be filtered here.